atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/CommitTsCacheTest.java

```diff
@@ -109,10 +111,41 @@ public void warmingCacheShouldNotPlaceUndueLoadOnTransactionService() throws Exc
             return timestamps.stream().collect(Collectors.toMap(n -> n, n -> n));
         }).when(mockTransactionService).get(any());
 
-        TLongSet initialTimestamps = new TLongHashSet();
-        initialTimestamps.addAll(LongStream.range(0, valuesToInsert).boxed().collect(Collectors.toList()));
+        Set<Long> initialTimestamps = LongStream.range(0, valuesToInsert).boxed().collect(Collectors.toSet());
+
+        alternativeLoader.loadBatch(initialTimestamps);
+        assertThat(alternativeLoader.load(valuesToInsert - 1)).isEqualTo(valuesToInsert - 1);
+    }
+
+    @Test
+    public void onlyRequestNonCachedTimestamps() throws Exception {
+        CommitTsCache alternativeLoader = CommitTsCache.create(mockTransactionService);
+
+        Set<Long> initialTimestamps = LongStream.range(0L, 20L).boxed().collect(Collectors.toSet());
+        doAnswer(invocation -> assertRequestedTimestampsAndMapIdentity(invocation, initialTimestamps))
+                .when(mockTransactionService).get(any());
+
+        alternativeLoader.loadBatch(initialTimestamps);
+        assertThat(alternativeLoader.load(19L)).isEqualTo(19L);
+
+        Set<Long> moreTimestamps = LongStream.range(10L, 22L).boxed().collect(Collectors.toSet());
+        doAnswer(invocation -> assertRequestedTimestampsAndMapIdentity(invocation, ImmutableSet.of(20L, 21L)))
+                .when(mockTransactionService).get(any());
+
+        alternativeLoader.loadBatch(moreTimestamps);
+        assertThat(alternativeLoader.load(21L)).isEqualTo(21L);
+
+        Set<Long> evenMoreTimestamps = LongStream.range(15L, 24L).boxed().collect(Collectors.toSet());
+        doAnswer(invocation -> assertRequestedTimestampsAndMapIdentity(invocation, ImmutableSet.of(22L, 23L)))
+                .when(mockTransactionService).get(any());
+
+        alternativeLoader.loadBatch(evenMoreTimestamps);
+        assertThat(alternativeLoader.load(23L)).isEqualTo(23L);
```

Comment:
OK. I guess you could use `Sets.difference()` rather than hand-computing the sets, though it's pretty obvious that those are the new ones that need to be requested